File: apps/web/src/lib/moves.ts
import type { Mat4, NodeMove, SceneGraph, SceneNode } from "@/types/contracts";

export type MoveSet = Record<string, NodeMove>;

/** Mirrors Lane C's move_node: turn about the node's own centre, then translate on the floor. */
export function moveNode(node: SceneNode, move: NodeMove): SceneNode {
  const m = node.transform.m;
  const angle = Math.atan2(m[4], m[0]) + (move.delta_rotation_z_degrees * Math.PI) / 180;
  const [cos, sin] = [Math.cos(angle), Math.sin(angle)];
  const x = m[3] + move.delta_translation.x;
  const y = m[7] + move.delta_translation.y;
  const z = m[11] + move.delta_translation.z;
  const transform = { m: [cos, -sin, 0, x, sin, cos, 0, y, 0, 0, 1, z, 0, 0, 0, 1] } as Mat4;
  return { ...node, transform };
}

/** Mirrors Lane C's RESTING_GAP: an underside higher than this above the floor was resting on something. */
const RESTING_GAP = 0.12;

function floorHeight(scene: SceneGraph): number {
  return scene.nodes.find((node) => node.kind === "floor")?.transform.m[11] ?? 0;
}

function underside(node: SceneNode): number {
  return node.transform.m[11] - node.dimensions.z / 2;
}

function restsOnSomething(node: SceneNode, floorZ: number): boolean {
  return node.kind === "object" && underside(node) > floorZ + RESTING_GAP;
}

function coversPoint(node: SceneNode, x: number, y: number): boolean {
  const m = node.transform.m;
  const [cos, sin] = [m[0], m[4]];
  const [dx, dy] = [x - m[3], y - m[7]];
  const localX = dx * cos + dy * sin;
  const localY = -dx * sin + dy * cos;
  return Math.abs(localX) <= node.dimensions.x / 2 && Math.abs(localY) <= node.dimensions.y / 2;
}

/** Mirrors Lane C's settle: sit on the highest top under the piece's centre, or on the floor. */
function settle(scene: SceneGraph, node: SceneNode, floorZ: number): SceneNode {
  const [x, y] = [node.transform.m[3], node.transform.m[7]];
  const surface = scene.nodes
    .filter((other) => other.id !== node.id && other.kind === "object" && coversPoint(other, x, y))
    .reduce((highest, other) => Math.max(highest, other.transform.m[11] + other.dimensions.z / 2), floorZ);
  const m = [...node.transform.m];
  m[11] = surface + node.dimensions.z / 2;
  return { ...node, transform: { m } as Mat4 };
}

export function applyMoves(scene: SceneGraph, moves: MoveSet): SceneGraph {
  if (Object.keys(moves).length === 0) return scene;
  const floorZ = floorHeight(scene);
  const resting = new Set(scene.nodes.filter((node) => moves[node.id] && restsOnSomething(node, floorZ)).map((node) => node.id));
  const moved = { ...scene, nodes: scene.nodes.map((node) => (moves[node.id] ? moveNode(node, moves[node.id]) : node)) };
  return { ...moved, nodes: moved.nodes.map((node) => (resting.has(node.id) ? settle(moved, node, floorZ) : node)) };
}

function sameEnvelope(node: SceneNode, next: SceneNode | undefined): next is SceneNode {
  return !!next && next.kind === node.kind && next.movable === node.movable &&
    (["x", "y", "z"] as const).every((axis) => node.dimensions[axis] === next.dimensions[axis]);
}

function floorMove(node: SceneNode, next: SceneNode): NodeMove | null {
  if (!node.movable || node.kind !== "object") return null;
  const m = node.transform.m, n = next.transform.m;
  const angle = Math.atan2(n[4], n[0]) - Math.atan2(m[4], m[0]);
  const move: NodeMove = {
    node_id: node.id,
    delta_translation: { x: n[3] - m[3], y: n[7] - m[7], z: 0 },
    delta_rotation_z_degrees: Math.atan2(Math.sin(angle), Math.cos(angle)) * 180 / Math.PI,
  };
  const HEIGHT = 11;
  return moveNode(node, move).transform.m.every((value, index) => index === HEIGHT || Math.abs(value - n[index]) < 1e-6) ? move : null;
}

function sameHeights(settled: SceneGraph, candidate: SceneGraph): boolean {
  const heights = new Map(candidate.nodes.map((node) => [node.id, node.transform.m[11]]));
  return settled.nodes.every((node) => Math.abs(node.transform.m[11] - (heights.get(node.id) ?? Number.NaN)) < 1e-6);
}

function sameRoom(base: SceneGraph, candidate: SceneGraph): boolean {
  return base.scan_id === candidate.scan_id && base.nodes.length === candidate.nodes.length;
}

type MoveOutcome = NodeMove | "unchanged" | null;

function moveTo(node: SceneNode, next: SceneNode | undefined): MoveOutcome {
  if (!sameEnvelope(node, next)) return null;
  if (next.transform.m.some((value) => !Number.isFinite(value))) return null;
  if (node.transform.m.every((value, index) => Math.abs(value - next.transform.m[index]) < 1e-6)) return "unchanged";
  return floorMove(node, next);
}

function movesBetween(base: SceneGraph, candidate: SceneGraph): NodeMove[] | null {
  const proposed = new Map(candidate.nodes.map((node) => [node.id, node]));
  if (proposed.size !== base.nodes.length) return null;
  const moves: NodeMove[] = [];
  for (const node of base.nodes) {
    const outcome = moveTo(node, proposed.get(node.id));
    if (outcome === null) return null;
    if (outcome !== "unchanged") moves.push(outcome);
  }
  return moves;
}

/** Recover only floor moves from a screened candidate for the existing review/save flow. */
export function candidateMoves(base: SceneGraph, candidate: SceneGraph): NodeMove[] | null {
  if (!sameRoom(base, candidate)) return null;
  const moves = movesBetween(base, candidate);
  if (!moves) return null;
  return sameHeights(applyMoves(base, Object.fromEntries(moves.map((move) => [move.node_id, move]))), candidate) ? moves : null;
}

export function withMove(moves: MoveSet, nodeId: string, dx: number, dy: number, degrees: number): MoveSet {
  const current = moves[nodeId] ?? { node_id: nodeId, delta_translation: { x: 0, y: 0, z: 0 }, delta_rotation_z_degrees: 0 };
  return {
    ...moves,
    [nodeId]: {
      node_id: nodeId,
      delta_translation: { x: current.delta_translation.x + dx, y: current.delta_translation.y + dy, z: 0 },
      delta_rotation_z_degrees: (current.delta_rotation_z_degrees + degrees) % 360,
    },
  };
}

export const METERS_PER_INCH = 0.0254;

File: packages/agents/standardphysics_agents/fix/moves.py
"""Applying a proposal to a layout.

A move translates across the room and turns about Z. Furniture standing on the
floor keeps its height. A piece that was resting on something else, a laptop on
a desk or a pillow on a sofa, settles onto whatever is under it where it lands,
or onto the floor when nothing is. Nothing here can change a dimension, because
there is no code path that writes one: the candidate node is copied from the
original with a new transform and nothing else.
"""

from __future__ import annotations

import math
from uuid import UUID

from standardphysics_contracts import Mat4, NodeMove, SceneGraph, SceneNode, Vec3, bounds_the_room, lies_flat
from standardphysics_pipeline import contains_point, footprint
from standardphysics_pipeline.footprints import rotation_about_z

RESTING_GAP = 0.12
"""A piece whose underside is more than this above the floor was resting on something."""


def _turned(node: SceneNode, degrees: float, position: Vec3) -> Mat4:
    """A transform holding the node's rotation about Z plus `degrees`.

    Furniture standing on a floor is a Z rotation and a translation, which is
    what RoomPlan reports and what a drag produces. A transform carrying
    anything else is not something a rearrangement should be composing with.
    """
    cos_t, sin_t = rotation_about_z(node)
    cos_d, sin_d = math.cos(math.radians(degrees)), math.sin(math.radians(degrees))
    cos_new = cos_t * cos_d - sin_t * sin_d
    sin_new = sin_t * cos_d + cos_t * sin_d
    return Mat4(
        m=[
            cos_new, -sin_new, 0.0, position.x,
            sin_new, cos_new, 0.0, position.y,
            0.0, 0.0, 1.0, position.z,
            0.0, 0.0, 0.0, 1.0,
        ]
    )


def move_node(node: SceneNode, move: NodeMove) -> SceneNode:
    origin = node.transform.position
    moved_to = Vec3(
        x=origin.x + move.delta_translation.x,
        y=origin.y + move.delta_translation.y,
        z=origin.z + move.delta_translation.z,
    )
    return node.model_copy(
        update={"transform": _turned(node, move.delta_rotation_z_degrees, moved_to)}
    )


def floor_height(graph: SceneGraph) -> float:
    floor = next((node for node in graph.nodes if lies_flat(node)), None)
    return floor.transform.position.z if floor else 0.0


def underside(node: SceneNode) -> float:
    return node.transform.position.z - node.dimensions.z / 2


def top_of(node: SceneNode) -> float:
    return node.transform.position.z + node.dimensions.z / 2


def rests_on_something(node: SceneNode, floor_z: float) -> bool:
    return not bounds_the_room(node) and underside(node) > floor_z + RESTING_GAP


def _surface_under(graph: SceneGraph, node: SceneNode, floor_z: float) -> float:
    """The highest top among the pieces directly under this one's centre."""
    centre = (node.transform.position.x, node.transform.position.y)
    tops = [
        top_of(other)
        for other in graph.nodes
        if other.id != node.id and not bounds_the_room(other) and contains_point(footprint(other), centre)
    ]
    return max([floor_z, *tops])


def settle(graph: SceneGraph, node: SceneNode, floor_z: float) -> SceneNode:
    """The node lowered or raised so its underside sits on the surface below it."""
    position = node.transform.position
    resting_at = _surface_under(graph, node, floor_z) + node.dimensions.z / 2
    return node.model_copy(
        update={"transform": _turned(node, 0.0, Vec3(x=position.x, y=position.y, z=resting_at))}
    )


def apply_moves(graph: SceneGraph, moves: list[NodeMove]) -> SceneGraph:
    """A new layout. The original is never touched, so a rejected candidate
    cannot leave anything behind."""
    by_node: dict[UUID, NodeMove] = {move.node_id: move for move in moves}
    floor_z = floor_height(graph)
    resting = {node.id for node in graph.nodes if node.id in by_node and rests_on_something(node, floor_z)}
    moved = graph.model_copy(
        update={"nodes": [move_node(node, by_node[node.id]) if node.id in by_node else node for node in graph.nodes]}
    )
    nodes = [settle(moved, node, floor_z) if node.id in resting else node for node in moved.nodes]
    return graph.model_copy(
        update={"nodes": nodes, "revision": graph.revision + 1, "base_hash": None}
    )


def without(graph: SceneGraph, node_ids) -> SceneGraph:
    """A layout with something set aside, for testing a relaxation only.

    This is never a proposal. Inventory is not adjustable, so removing a chair
    is a question for the owner and this exists to find out whether asking it
    would even help.
    """
    removed = set(node_ids)
    return graph.model_copy(
        update={"nodes": [n for n in graph.nodes if n.id not in removed]}
    )


def unlocked(graph: SceneGraph, node_ids) -> SceneGraph:
    """A layout where something fixed became movable, for testing a relaxation."""
    targets = set(node_ids)
    return graph.model_copy(
        update={
            "nodes": [
                node.model_copy(update={"movable": True})
                if node.id in targets
                else node
                for node in graph.nodes
            ]
        }
    )


def footprint_span(node: SceneNode, axis: tuple[float, float]) -> float:
    """How far the node reaches along `axis`, from its centre."""
    centre = node.transform.position
    return max(
        abs((x - centre.x) * axis[0] + (y - centre.y) * axis[1])
        for x, y in footprint(node)
    )
